File: src/services/ai.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  AIProviderConfig,
  AIContext,
  AISuggestionResponse,
  AIConnectionTestResult,
  AISettings,
  AISuggestionRequest,
  AIProviderType,
  CreateAIProviderRequest,
  UpdateAIProviderRequest,
} from "../types/ai";

/**
 * AI Service
 * Handles interaction with the Rust backend for AI functionality
 */
export const aiService = {
  /**
   * Get global AI settings
   */
  async getSettings(): Promise<AISettings> {
    return invoke("get_ai_settings");
  },

  /**
   * Update global AI settings
   */
  async updateSettings(settings: AISettings): Promise<void> {
    await invoke("update_ai_settings", { settings });
  },

  /**
   * Get all configured AI providers
   */
  async getProviders(): Promise<AIProviderConfig[]> {
    return invoke("get_ai_providers");
  },

  /**
   * Add a new AI provider
   */
  async addProvider(
    request: CreateAIProviderRequest
  ): Promise<AIProviderConfig> {
    return invoke("add_ai_provider", { request });
  },

  /**
   * Update an existing AI provider
   */
  async updateProvider(
    id: string,
    request: UpdateAIProviderRequest
  ): Promise<AIProviderConfig> {
    return invoke("update_ai_provider", { id, request });
  },

  /**
   * Remove an AI provider
   */
  async removeProvider(id: string): Promise<void> {
    await invoke("remove_ai_provider", { id });
  },

  /**
   * Test connection to an AI provider
   */
  async testConnection(
    config: AIProviderConfig
  ): Promise<AIConnectionTestResult> {
    return invoke("test_ai_connection", { config });
  },

  /**
   * Get command suggestions
   */
  async getSuggestions(
    request: AISuggestionRequest
  ): Promise<AISuggestionResponse> {
    return invoke("get_ai_suggestions", { request });
  },

  /**
   * Get available models for a provider type (Static helper)
   */
  getAvailableModels(type: AIProviderType): string[] {
    switch (type) {
      case "openai":
        return [
          "gpt-4o",
          "gpt-4o-mini",
          "gpt-4-turbo",
          "gpt-4",
          "gpt-3.5-turbo",
        ];
      case "anthropic":
        return [
          "claude-3-5-sonnet-latest",
          "claude-3-5-haiku-latest",
          "claude-3-opus-latest",
        ];
      case "gemini":
        return ["gemini-2.0-flash", "gemini-1.5-flash", "gemini-1.5-pro"];
      case "ollama":
        return [
          "llama3.2",
          "llama3.1",
          "mistral",
          "codellama",
          "deepseek-coder",
        ];
      case "custom":
        return ["gpt-3.5-turbo"];
      default:
        return [];
    }
  },
};

File: src-tauri/src/commands/ai.rs
use crate::error::AppError;
use crate::models::ai::{
    AIConnectionTestResult, AIProviderConfig, AISettings, AISuggestionRequest,
    AISuggestionResponse, CreateAIProviderRequest, UpdateAIProviderRequest,
};
use crate::state::AppState;

#[tauri::command]
pub async fn get_ai_settings(state: tauri::State<'_, AppState>) -> Result<AISettings, AppError> {
    Ok(state.ai_service.get_settings().await)
}

#[tauri::command]
pub async fn update_ai_settings(
    state: tauri::State<'_, AppState>,
    settings: AISettings,
) -> Result<(), AppError> {
    state.ai_service.update_settings(settings).await
}

#[tauri::command]
pub async fn get_ai_providers(
    state: tauri::State<'_, AppState>,
) -> Result<Vec<AIProviderConfig>, AppError> {
    Ok(state.ai_service.get_providers().await)
}

#[tauri::command]
pub async fn add_ai_provider(
    state: tauri::State<'_, AppState>,
    request: CreateAIProviderRequest,
) -> Result<AIProviderConfig, AppError> {
    state.ai_service.add_provider(request).await
}

#[tauri::command]
pub async fn update_ai_provider(
    state: tauri::State<'_, AppState>,
    id: String,
    request: UpdateAIProviderRequest,
) -> Result<AIProviderConfig, AppError> {
    state.ai_service.update_provider(id, request).await
}

#[tauri::command]
pub async fn remove_ai_provider(
    state: tauri::State<'_, AppState>,
    id: String,
) -> Result<(), AppError> {
    state.ai_service.remove_provider(id).await
}

#[tauri::command]
pub async fn test_ai_connection(
    state: tauri::State<'_, AppState>,
    config: AIProviderConfig,
) -> Result<AIConnectionTestResult, AppError> {
    state.ai_service.test_connection(config).await
}

#[tauri::command]
pub async fn get_ai_suggestions(
    state: tauri::State<'_, AppState>,
    request: AISuggestionRequest,
) -> Result<AISuggestionResponse, AppError> {
    state.ai_service.get_suggestions(request).await
}
